p2p/test/muxer-negotiation/muxer_negotiation_test.go

```diff
@@ -0,0 +1,199 @@
+			cliTrans:      []network.Multiplexer{mplex.DefaultTransport, yamux.DefaultTransport},
+			expectedMuxer: "/yamux/1.0.0"},
+		{svrMuxers: []string{"/mplex/6.7.0", "/yamux/1.0.0"},
+			svrTrans:      []network.Multiplexer{mplex.DefaultTransport, yamux.DefaultTransport},
+			cliMuxers:     []string{"/yamux/1.0.0", "/mplex/6.7.0"},
+			cliTrans:      []network.Multiplexer{yamux.DefaultTransport, mplex.DefaultTransport},
+			expectedMuxer: "/mplex/6.7.0"},
+	}
+
+	doMuxerNegotiation := func(t *testing.T, secType string, svrMuxers, cliMuxers []string, svrTrans, cliTrans []network.Multiplexer, expected string) {
+		sctx, sCancel := context.WithCancel(context.Background())
+		cctx, cCancel := context.WithCancel(context.Background())
+
+		sSec, svrh, err := makeHost(secType, svrMuxers, svrTrans, 58568)
+		require.NoError(t, err)
+		require.NotNil(t, sSec)
+		require.NotNil(t, svrh)
+
+		cSec, clih, err := makeHost(secType, cliMuxers, cliTrans, 59569)
+		require.NoError(t, err)
+		require.NotNil(t, cSec)
+		require.NotNil(t, clih)
+
+		ready := make(chan struct{})
+		go func() {
+			runServer(sctx, svrh)
+			close(ready)
+			<-sctx.Done()
+		}()
+
+		<-ready
+		runClientAndCheckMuxer(cctx, clih, getHostAddress(svrh), cSec, sSec, expected, t)
+		clih.Close()
+		svrh.Close()
+		cCancel()
+		sCancel()
+	}
+
+	golog.SetAllLoggers(golog.LevelInfo)
+
+	for _, secType := range secTypes {
+		for i, testCase := range testCases {
+			testName := "Test muxer negotiation for " + secType + ", case " + fmt.Sprint(i)
+			t.Run(testName, func(t *testing.T) {
+				doMuxerNegotiation(t, secType, testCase.svrMuxers, testCase.cliMuxers, testCase.svrTrans, testCase.cliTrans, testCase.expectedMuxer)
+			})
+		}
+	}
+}
+
+func makeHost(transportType string, muxers []string, muxTrans []network.Multiplexer, port int) (*TransportWithMuxer, host.Host, error) {
+	r := rand.Reader
+
+	priv, _, err := crypto.GenerateKeyPairWithReader(crypto.RSA, 2048, r)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	secTrans, err := New(priv, muxers, transportType)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	opts := []libp2p.Option{
+		libp2p.ListenAddrStrings(fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", port)),
+		//libp2p.DefaultListenAddrs,
+		libp2p.Identity(priv),
+		libp2p.Security(transportType, secTrans),
+	}
+	for i := 0; i < len(muxers); i++ {
+		opts = append(opts, libp2p.Muxer(muxers[i], muxTrans[i]))
+	}
+	h, err := libp2p.New(opts...)
+	return secTrans, h, err
+}
+
+func getHostAddress(ha host.Host) string {
+	// Build host multiaddress
+	hostAddr, _ := ma.NewMultiaddr(fmt.Sprintf("/p2p/%s", ha.ID().Pretty()))
+
+	addr := ha.Addrs()[0]
+	return addr.Encapsulate(hostAddr).String()
+}
+
+func runServer(ctx context.Context, ha host.Host) {
+	ha.SetStreamHandler("/echo/1.0.0", streamHandler)
+}
+
+func runClientAndCheckMuxer(ctx context.Context, ha host.Host, targetPeer string, cSec, sSec *TransportWithMuxer, expected string, t *testing.T) {
```

Comment:
1. You don't need `getHostAddress` if you pass a `peer.AddrInfo` to this function.
2. By convention, `*testing.T` should be the first parameter.

Reply:
1. Not sure I am exactly following, but yes we don't have to convert it to string and back to addinfo, if that is what you mean. Updated the code.
2. yes, changed code.

Reply:
1. Not sure I am exactly following, but yes we don't have to convert it to string and back to addinfo, if that is what you mean. Updated the code.
2. yes, changed code.